### ISSUE TITLE
Differentiation doesn't behave correctly

### ISSUE THREAD

**Issue by yolandalalala** [uclnlp/torch-imle#4]
I'm working on some discrete combinatorial optimization problems, in which IMLE seems to be highly suitable. However, in some instances, it does not behave as anticipated.
Here's a simplified example for better understanding:
### Definition of Solver
```python
@imle(target_distribution=TargetDistribution(alpha=1.0, beta=1.0),
      noise_distribution=SumOfGammaNoiseDistribution(k=10, nb_iterations=100),
      input_noise_temperature=10.0,
      target_noise_temperature=10.0)
def solver(weights_batch: Tensor) -> Tensor:
    """
    weights_batch: is a tensor of float values range between -1 and 1
    """
    return (weights_batch > 0).float()
```
### Definition of Parameters, Optimizer, and Labels
```python
param = nn.Parameter(torch.zeros(8, requires_grad=True))
optimizer = torch.optim.AdamW([param], lr=0.1)
y1 = torch.tensor([0., 0., 0., 0., 1., 1., 1., 1.])
y2 = torch.tensor([1., 1., 1., 1., 0., 0., 0., 0.])
```
### Train with `y1` as labels
```python
for i in range(100):
    w = torch.tanh(param)
    pred = solver(w)
    loss = (pred - y1).square().sum()
    loss.backward()
    optimizer.step()
param
```
#### Expected Result:
- Values in `param[:4]` should all turn negative after training.
- Values in `param[4:]` should all turn positive after training.
#### Actual Result: as expected
```
Parameter containing:
tensor([ -9.9164, -10.1681,  -9.8578,  -9.9074,  10.4283,   9.9732,  10.2848,
         10.2281], requires_grad=True)
```
### Train with `y2` as labels
Continue training with negated labels without resetting the parameters.
```python
for i in range(100):
    w = torch.tanh(param)
    pred = solver(w)
    loss = (pred - y2).square().sum()
    loss.backward()
    optimizer.step()
param
```
#### Expected Result:
- The values in param[:4] should start to rise and eventually turn positive.
- The values in param[4:] should start to fall and eventually turn negative.
#### Actual Result:
- `param[:4]` continues to fall, while `param[4:]` continues to rise.
```
Parameter containing:
tensor([-18.7772, -19.0104, -18.5812, -18.6452,  19.3325,  18.8648,  19.1460,
         19.0642], requires_grad=True)
```
I've attempted this with various hyperparameter configurations, but the outcome remains the same. I'm uncertain as to where the potential problems might lie. Could you provide some insights?

**Comment by mniepert** [comment]
Thanks for your question. Which hyperparameters have you played with?

IMLE executes the solver twice, once with the input w and once with an updated w' which is w moved in the direction of the negative gradient. One way to check what's going on is to print the solver output for these two invocations. If you paste some example vectors, we can see if the gradients from IMLE are what is to be expected. 

Generally, my suggestion is to start with the standard Gumbel noise.

**Comment by yolandalalala** [comment]
> Which hyperparameters have you played with?

I played with target distribution `alpha` and `beta`, noise distribution `k`, and noise temperatures.

> One way to check what's going on is to print the solver output for these two invocations.

I printed the solver outputs when using `y2` as labels for 20 iterations. 
```
tensor([1., 0., 0., 0., 1., 1., 1., 1.])
tensor([1., 1., 0., 0., 1., 1., 1., 0.])
----
tensor([0., 0., 0., 0., 1., 1., 1., 1.])
tensor([0., 1., 1., 0., 0., 0., 1., 0.])
----
tensor([0., 0., 0., 1., 1., 1., 1., 1.])
tensor([1., 1., 1., 1., 1., 1., 1., 0.])
----
tensor([0., 0., 0., 0., 1., 1., 0., 1.])
tensor([1., 0., 1., 1., 0., 0., 0., 0.])
----
tensor([0., 1., 0., 0., 1., 1., 1., 1.])
tensor([1., 1., 1., 1., 0., 0., 1., 0.])
----
tensor([0., 0., 0., 0., 0., 1., 0., 1.])
tensor([0., 1., 1., 1., 0., 0., 0., 1.])
----
tensor([0., 0., 0., 0., 1., 1., 1., 1.])
tensor([1., 1., 1., 1., 0., 0., 1., 0.])
----
tensor([1., 0., 1., 0., 1., 1., 1., 1.])
tensor([1., 1., 1., 1., 1., 0., 0., 0.])
----
tensor([0., 0., 1., 1., 0., 0., 1., 0.])
tensor([1., 1., 1., 1., 0., 0., 1., 0.])
----
tensor([0., 1., 0., 0., 1., 1., 1., 1.])
tensor([1., 1., 0., 1., 0., 0., 1., 0.])
----
tensor([0., 0., 0., 0., 1., 1., 0., 1.])
tensor([1., 1., 1., 1., 0., 0., 0., 0.])
----
tensor([0., 0., 1., 0., 1., 1., 1., 1.])
tensor([1., 1., 1., 1., 1., 0., 0., 0.])
----
tensor([0., 0., 0., 0., 0., 1., 1., 1.])
tensor([0., 1., 1., 0., 0., 0., 0., 0.])
----
tensor([0., 0., 0., 0., 1., 0., 1., 1.])
tensor([0., 1., 1., 0., 0., 0., 0., 0.])
----
tensor([0., 1., 1., 1., 1., 1., 1., 1.])
tensor([1., 1., 1., 1., 0., 0., 0., 1.])
----
tensor([0., 1., 0., 0., 1., 1., 1., 1.])
tensor([0., 1., 1., 0., 0., 0., 0., 0.])
----
tensor([0., 0., 1., 0., 1., 1., 0., 1.])
tensor([1., 1., 1., 1., 0., 0., 0., 0.])
----
tensor([0., 1., 0., 1., 0., 0., 0., 1.])
tensor([1., 1., 0., 1., 0., 0., 0., 1.])
----
tensor([1., 0., 0., 0., 1., 1., 1., 0.])
tensor([1., 1., 1., 1., 1., 0., 1., 0.])
----
tensor([0., 0., 0., 0., 0., 1., 1., 1.])
tensor([0., 1., 1., 1., 0., 0., 1., 0.])
----

Parameter containing:
tensor([-11.7657, -11.4689, -11.7747, -11.8080,  11.8574,  11.9253,  11.8690,
         11.7945], requires_grad=True)
```

> Generally, my suggestion is to start with the standard Gumbel noise.

In this repo, I only see `SumOfGammaNoiseDistribution` is implemented. Do I need to implement my own standard Gumbel noise distribution?

Thanks!

**Comment by mniepert** [comment]
Thanks. It’s for sure a problem with the noise distribution. The magnitude of the noise is too strong which shows in your rather random vectors. If you set k and the noise temperature to a lower value (eg all 1) it should work. 

There is a newer repo (for adaptive IMLE) which also has the Gumbel noise implemented. You can see it here: https://github.com/EdinburghNLP/torch-adaptive-imle/blob/main/imle/noise.py

**Comment by mniepert** [comment]
Actually, I think there is more to it:

For instance,  the gradients that IMLE computes for the first 4 variables are correct (mostly -1 and sometimes 0). -1 means „increase“ the value. So there is also a strong possibility that the tanh is completely saturated and gives zero gradients. 

You could remove the tanh to check this

**Comment by mniepert** [comment]
My point is: IMLE operates and returns gradients as expected even with the current noise setting. There must be something else at play here that has nothing to do with IMLE (eg zero gradients from tanh, underflow from tanh, adaptive optimizer went off the rails due to seeing gradients in a single direction for many steps which suddenly change direction etc)

**Comment by yolandalalala** [comment]
I appreciate the comprehensive analysis and the detailed insights provided. After conducting some additional debugging, I have discovered that the issue was due to the gradient not being cleared after each iteration, while IMLE is functioning as expected per se. I apologize for the confusion caused. Given these findings, I will now be closing this issue.
